src/FirmwarePlugin/FirmwarePlugin.cc

```diff
@@ -663,3 +667,73 @@ uint32_t FirmwarePlugin::highLatencyCustomModeTo32Bits(uint16_t hlCustomMode)
     return hlCustomMode;
 }
 
+void FirmwarePlugin::checkIfIsLatestStable(Vehicle* vehicle)
+{
+    // This is required as mocklink uses a hardcoded firmware version
+    if (qgcApp()->runningUnitTests()) {
+        qCDebug(FirmwarePluginLog) << "Skipping version check";
+        return;
+    }
+    QString versionFile = _getLatestVersionFileUrl(vehicle);
+    qCDebug(FirmwarePluginLog) << "Downloading" << versionFile;
+    QGCFileDownload* downloader = new QGCFileDownload(this);
+    connect(
+        downloader,
+        &QGCFileDownload::downloadFinished,
+        this,
+        [vehicle, this](QString remoteFile, QString localFile) {
+            _versionFileDownloadFinished(remoteFile, localFile, vehicle);
+            sender()->deleteLater();
+        });
+    downloader->download(versionFile);
+}
+
+void FirmwarePlugin::_versionFileDownloadFinished(QString& remoteFile, QString& localFile, Vehicle* vehicle)
```

Comment:
I think the implementation here needs to be moved back to ardupilotfirmwareplugin, and a new one made for px4. 

For PX4, as far as I can see, there is no guarantee that the first regex match will be the stable version number. It might be a beta release, prerelease, or old release. While none of these things are true at the moment, there aren't any guarantees that it will remain this way.

Reply:
That's why the regex is looking for "vX.X.X Stable". All stable versions since 1.5.5 (2017-01-25) in [the file](https://api.github.com/repos/PX4/Firmware/releases) follow this naming.  But I agree this is not as robust.

At [FirmwareUpgradeController.cc](https://github.com/mavlink/qgroundcontrol/blob/master/src/VehicleSetup/FirmwareUpgradeController.cc#L848) the json is properly processed, but then it is also iterated from top to bottom until it finds the first beta (prerelease=true) and first stable release (prerelease=false).

The current way there's less code, but might break if they change the naming standard.

Reply:
My concern is also the github api returning the list in chronological order. https://developer.github.com/v3/repos/releases/#list-releases-for-a-repository

It will probably continue to work as-is. We can wait till to fix it if it breaks.

Reply:
I'm slightly uncomfortable with this not being broken out to PX4 and ArduPilot versions. Mainly because the _versionRegEx mechanism is not super clear at first glance. But that said it works, so fine.